stellar/controllers/stellar_get_cancel_transaction.go

```diff
@@ -0,0 +1,47 @@
+package controllers
+
+import (
+	"database/sql"
+	"net/http"
+
+	"github.com/stellar/go/support/log"
+	"github.com/stellar/starbridge/store"
+)
+
+type StellarGetCancelTransactionForEthereum struct {
+	Store *store.DB
+}
+
+func (c *StellarGetCancelTransactionForEthereum) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	ethereumTransactionHash := r.PostFormValue("transaction_hash")
+
+	// Ensure incoming transaction exists
+	_, err := c.Store.GetIncomingEthereumTransactionByHash(r.Context(), ethereumTransactionHash)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			w.WriteHeader(http.StatusNotFound)
+		} else {
+			log.WithField("error", err).Error("Error getting an incomming ethereum transaction")
+			w.WriteHeader(http.StatusInternalServerError)
+		}
+		return
+	}
+
+	// Check if outgoing transaction exists
+	outgoingTransaction, err := c.Store.GetOutgoingStellarTransactionForEthereumByHash(r.Context(), ethereumTransactionHash)
+	if err != nil && err != sql.ErrNoRows {
+		log.WithField("error", err).Error("Error getting an outgoing stellar transaction for ethereum")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	// Error if outgoing tx exists and is pending or successful
```

Comment:
what if we receive a SignatureRequest and a StellarGetCancelTransactionForEthereum request at the same time ? I think there's a race condition both requests are considered valid and we generate both a withdrawal and a refund

Reply:
The signing requests are processed in a single go routine in a backend worker.